support for Spring Cloud properties with DB2 services i.e. cloud.services.{service-name}.connection.{property}

mariobriggs opened the pull request:
Use-case : Pushing a Play framework generated .zip file to IBM Bluemix (CloudFoundry based) with the [java-buildpack](https://github.com/cloudfoundry/java-buildpack). The application has a bound DB2 service instance. The [app.conf](https://github.com/cloudfoundry-samples/zentasks-scala-cloudfoundry/blob/master/conf/application.conf) references the cloud.services.xxx.connection.uri. Need these Spring Cloud {cloud.services...) variables to be populated from the bound DB2 service.

Attached PR adds the DB2ServiceInfoCreator & registers it. Added CloudFoundryConnectorDB2ServiceTest and run it.  2 test cases fail with 'WARNING: No suitable service info creator found for service db2-ups Did you forget to add a ServiceInfoCreator?' , but i get the same failures even with Oracle, Postgres & MySQL tests. Screen shot attached
![pgcloudservices](https://cloud.githubusercontent.com/assets/3006502/7806258/62881fa6-039c-11e5-8d48-7cd635a1cdb3.png)

Also is there a way i can test on CloudFoundry(Bluemix) with the jars from my fork of the repo ?

scottfrederick commented on spring-cloud-core/src/main/java/org/springframework/cloud/service/common/DB2ServiceInfo.java:
```diff
@@ -0,0 +1,27 @@
+			return getUriInfo().getUriString();
+		}
+		                    
+		return String.format("jdbc:%s://%s:%d/%s",
```
The crated JDBC URL does include the username and password from credentials. It seems this would be needed for authenticated database connections.

mariobriggs replied:
@scottfrederick DB2 JDBC driver does not support passing the username and password in the URL (either like jdbc:db2://username:password@... or jdbc:db2://.../dbname?user=...&password=...). It is required to pass it as a separate properties or separate string arguments to the DriverManager.connect(). This is why i didn't include them in the JDBC URL

mariobriggs replied:
@scottfrederick  i just found out that it supports in the format jdbc:db2://.../dbname;user=...;password=...;). So ignore my above comment and i will update the PR with this change

scottfrederick commented:
@mariobriggs The `'WARNING: No suitable service info creator found` messages you are seeing are normal. There are tests that force this condition to validate that the correct exceptions are thrown from the library, which also causes that message to be send to stderr. That's not the source of the test failure. It looks like the true failure is the `java.lang.ClassCastException`, but I don't know why that is happening. 

When I build your branch all tests succeed for me. Can you try running the tests from the command line (e.g. `./gradlew clean test`) to see if this is a problem with Eclipse? 

FYI, @gautierj previously submitted a PR (#87) for DB2 support, but it had a few minor formatting issues that I was waiting on before merging. These two PRs are very similar, but Jacques' includes a few details that are missing here, like the `DB2DataSourceCreator.java` that will be useful when using this DB2 support in Spring applications. It would be good to compare these two PRs and make sure everything is covered. 

> Also is there a way i can test on CloudFoundry(Bluemix) with the jars from my fork of the repo ?

You should be able to build the library locally and install into your local maven repo (e.g. `./gradlew clean install`). Use the version `1.2.0.BUILD-SNAPSHOT` in a client application to get your locally built version. If you are using Gradle in your client app, make sure you have `mavenLocal()` in the `repositories` section of the build file along with `mavenCentral()`.

mariobriggs commented:
@scottfrederick  i have added the DB2DataSourceCreator and its respective tests. The tests went through fine. I have also updated the tags in DB2ServiceInfoCreator looking at @gautierj PR and also fixed the getJDBCUrl() specifically for DB2. 

The class exception is expected after the 'no suitable service info creator' warning since the class returned is BaseServiceInfo and not the required Postgres/Oracle/MySql ServiceInfo class. I can try on the command line when i get some time. Since the those tests passed for you, i guess it is good to go.

mariobriggs commented:
@scottfrederick any update on this or u need something further from me ?

scottfrederick commented:
@mariobriggs Looks good, thanks!

mariobriggs commented:
@scottfrederick  thanks. Will this make it into v1.2.0.RC2 or v1.2.0 release ?

scottfrederick commented:
@mariobriggs I think we can include this in 1.2.0.RELEASE. The 1.2.0.RC was published mainly to let people test some API changes, but this addition of DB2 support is pretty safe. 

Have you had a chance to test the code as built into 1.2.0.SNAPSHOT (from [repo.spring.io/lib-snapshot](https://github.com/spring-projects/spring-framework/wiki/Downloading-Spring-artifacts#via-the-spring-repository)) and see if it works as merged?

mariobriggs commented:
@scottfrederick thanks for the info about the 1.2.0 SNAPSHOT (from repo.spring.io/lib-snapshot). I used that with [this sample](https://github.com/cloudfoundry-samples/hello-spring-cloud) and the Datasource was injected correctly to DB2 (yay). So the fix is good.

So then i moved to my original use-case - having the 'cloud.services.xxx.connection.uri' env variable injected. Using the same app mentioned above the env variable does not get injected. It is not getting injected even when i have a postgresql service... so it seems i might have to add something to the sample to exercise that functionality.

 Do u have any ideas how to test that part by adding to above sample app ? (the original play app doesnt push its own spring-cloud* jars and even by it adding it, didnt help.. I see that buildpack says 'Downloading Play Framework Auto Reconfiguration 1.7.0_RELEASE from https://download.run.pivotal.io/auto-reconfiguration/auto-reconfiguration-1.7.0_RELEASE.jar' and that jar contains the spring-cloud-* classes. So that way i might have to fork the buildpack and have it pull my from repo.spring.io/lib-snapshot as mentioned [here](https://github.com/cloudfoundry/java-buildpack/issues/179#issuecomment-115161908). But if there is a easier way, prefer that)

mariobriggs commented:
@scottfrederick please see PR #123 which is required to have the jdbcurl invoked on the DB2ServiceInfo class. Appreciate if you can merge

scottfrederick commented:
Thanks @mariobriggs, that PR is merged.

mariobriggs commented:
thank you
